content/blog/2018/app-metrics-istio/index.md

```diff
@@ -0,0 +1,82 @@
+twitter: meteatamel
+weight: 78
+keywords: [metrics, prometheus]
+---
+## Background
+
+The default metrics sent by Istio are useful to get an idea on how the traffic flows in your cluster. However, to understand how your application behaves, you also need application metrics. 
+
+[Prometheus](https://prometheus.io/) has [client libraries](https://prometheus.io/docs/instrumenting/clientlibs/) that you can use to instrument your application and send those metrics. This is good but it raises some questions: 
+* Where do you collect those metrics? 
+* Do you use Istio’s Prometheus or set up your own Prometheus? 
+* If you use Istio’s Prometheus, what configuration do you need to get those metrics scraped? 
+
+Let’s try to answer these questions.
+
+## Istio vs. standalone Prometheus
+In Prometheus, there’s a [federation](https://prometheus.io/docs/prometheus/latest/federation/) feature that allows a Prometheus server to scrape metrics from another Prometheus server. If you want to keep Istio metrics and application metrics separate, you can set up a seperate Prometheus server for application metrics. Then, you can use federation to scrape  those application metrics scraped with Istio’s Prometheus server. 
+
+A simpler approach is to scrape the application’s metrics  using Istio’s Prometheus directly and that’s what I want to focus on here. 
+
+## Sending application metrics
+To send custom metrics from your application, you need to instrument your application using Prometheus’ [client libraries](https://prometheus.io/docs/instrumenting/clientlibs/). Which  library to use depends on the language you’re using. As a C#/.NET developer, I used the  [.NET client](https://github.com/prometheus-net/prometheus-net) for Prometheus and [this blog post](https://www.olivercoding.com/2018-07-22-prometheus-dotnetcore/) from Daniel Oliver has step-by-step instructions on how to send custom metrics from an ASP.NET Core application and see them in a local Prometheus server.
+
+One thing you need to pay attention to is the port where you’re exposing your Prometheus metrics. In ASP.NET Core, the default port is **5000**. When running locally, application metrics are exposed on `localhost:5000/metrics`. However, when you containerize your app, it’s common to expose your application over a different port, like **8080**, and this becomes relevant later when we talk about configuration.
+
+Assuming that you containerized and deployed your application on an Istio-enabled Kubernetes cluster, let’s now take a look at what we need to do to get these application metrics scraped by Istio’s Prometheus.
+
+## Configuration
+In Istio 1.0.5, the default installation files for Kubernetes, `istio-demo.yaml` or `istio-demo-auth.yaml`, already have scraping configurations for Prometheus under a `ConfigMap`. You can just search for `prometheus.yml`. There are 2 scraping jobs that are relevant for application metrics: 
+
+{{< text bash >}}
+   - job_name: 'kubernetes-pods'
```

Comment:
Make sure that the first level of your code is on the same level as the `{{< text yaml >}}` tag to avoid linter issues.
Keep in mind that each level of indentation must be 4 spaces.
```suggestion
- job_name: 'kubernetes-pods'
```